Add factory for tester.state with recent fork-level

konradkonrad opened the pull request:
A common pitfall for developers is using vanilla `tester.state` with contracts
that use more recent opcodes (ending up with `invalid opcode` errors).

The factory function `latest_state`, prepares the state for any developer with
the latest fork block number (as set in ethereum.config.LATEST_FORK_BLKNUMBER).

This way, legacy applications won't change their behavior, but newer users/
developers can be encouraged to just use `latest_state()`.

konradkonrad committed:
Add factory for tester.state with recent fork-level

A common pitfall for developers is using vanilla `tester.state` with contracts
that use more recent opcodes (ending up with `invalid opcode` errors).

The factory function `latest_state`, prepares the state for any developer with
the latest fork block number (as set in ethereum.config.LATEST_FORK_BLKNUMBER).

This way, legacy applications won't change their behavior, but newer users/
developers can be encouraged to just use `latest_state()`.

Since inception of this PR, this got rebased on latest develop to now point to
the ANTI_DOS_FORK.

LefterisJP commented:
Simple and nice. Looks good to me. Let's see if the tests pass.

chfast commented:
@mfranciszkiewicz, can you review this?

mfranciszkiewicz commented on ethereum/tester.py:
```diff
@@ -416,3 +416,23 @@ def revert(self, data):
+    """
+    blockchain_state = state(**kwargs)
+
+    if blknum is not None:
```
I believe you meant `if blknum is None` here, otherwise `blockchain_state.block.number` is either `LATEST_APPLIED_FORK_BLKNUM + 1` or `None`.

konradkonrad replied:
:doh:
Thanks for spotting this! That was a pretty terrible `LOC/error` ratio :open_mouth: 
Fixed and added tests!

konradkonrad committed:
Fix blknum logic for latest_state and add tests

This fixes an error with the `blknum` parameter and changes
the logic to set the state to the actual `LATEST_APPLIED_FORK_BLKNUM` (instead
of `+ 1`), since rules are applied starting from the block.

Also added tests to ensure the intended behavior.

konradkonrad commented:
I don't think this will ever be merged...closing it!
